[threaded-animation-resolution] store accelerated effects and base values in the remote layer tree

graouts opened the pull request:
#### c4c680697bdba4208581dc7fd40ffdd202482c08
<pre>
[threaded-animation-resolution] store accelerated effects and base values in the remote layer tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=265880">https://bugs.webkit.org/show_bug.cgi?id=265880</a>

Reviewed by Simon Fraser.

While we correctly encode the `AcceleratedEffectStack` from WebCore to be sent over IPC to
the UI process whenever there is a meaningful change, we don&apos;t do anything with these effects
yet. This patch creates a `RemoteAcceleratedEffectStack` and store effects and base values on it
and lays the groundwork for the next patch for the threaded animation resolution feature where
we will schedule and resolve animations in the remote layer tree.

On iOS, the `RemoteAcceleratedEffectStack` is owned by a `RemoteLayerTreeNode` which will access
on the main thread for animation resolution.

On macOS, the `RemoteAcceleratedEffectStack` is owned by the `RemoteLayerTreeEventDispatcher` which
will allow access in a thread-safe manner for animation resolution.

* Source/WebCore/platform/animation/AcceleratedEffectStack.h:
(WebCore::AcceleratedEffectStack::primaryLayerEffects const): Deleted.
(WebCore::AcceleratedEffectStack::backdropLayerEffects const): Deleted.
(WebCore::AcceleratedEffectStack::baseValues): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffectStack.mm:
(WebCore::AcceleratedEffectStack::create):
(WebCore::AcceleratedEffectStack::AcceleratedEffectStack):
(WebCore::AcceleratedEffectStack::setEffects):
(WebCore::AcceleratedEffectStack::~AcceleratedEffectStack): Deleted.
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::setAcceleratedEffectsAndBaseValues):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm: Added.
(WebKit::RemoteAcceleratedEffectStack::create):
(WebKit::RemoteAcceleratedEffectStack::RemoteAcceleratedEffectStack):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationsWereAddedToNode):
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::threadedAnimationResolutionEnabled const):
(WebKit::RemoteLayerTreeHost::animationsWereAddedToNode):
(WebKit::RemoteLayerTreeHost::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::takeEffectStack):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxy::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::animationsWereAddedToNode):
(WebKit::RemoteLayerTreeEventDispatcher::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyMac::animationsWereRemovedFromNode):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Co-authored-by: Matt Woodrow &lt;mattwoodrow@apple.com&gt;
Canonical link: <a href="https://commits.webkit.org/271692@main">https://commits.webkit.org/271692@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d381934a5ebff692d0a51da5f621caebb293875

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29286 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7956 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30624 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31864 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29885 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10096 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5258 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29559 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6614 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/25076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5693 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/26123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33205 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26744 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26554 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3980 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7489 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3765 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -238,4 +243,32 @@
+    WebCore::AcceleratedEffects clonedEffects;
+    clonedEffects.reserveCapacity(effects.size());
+    for (auto& effect : effects)
+        clonedEffects.append(effect->clone());
```
Do we need the deep-clone here? `AcceleratedEffects` is a `Vector` of `Refs`, can we just copy the vector so that we get new strong refs to the same objects?

graouts replied:
We don't need to deep clone at the moment anyway. I can't remember why I added that originally, but I think it might have had something to do with blending. We'll revisit this when the time comes but the new patch no longer calls `clone()` on the effects.

mattwoodrow commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -238,4 +243,32 @@
+    if (effects.isEmpty())
+        return;
+
+    m_effects = new RemoteAcceleratedEffectStack(layer().bounds, host->acceleratedTimelineTimeOrigin());
```
I think this needs to have `adoptRef`.

graouts replied:
Yes, this should be behind `::create()` methods like `RefPtr` classes typically offer.

mattwoodrow commented on Source/WebCore/platform/graphics/GraphicsLayer.cpp:
```diff
@@ -764,7 +764,7 @@ void GraphicsLayer::setAcceleratedEffectsAndBaseValues(AcceleratedEffects&& effe
 
     if (!m_effectStack)
-        m_effectStack = makeUnique<AcceleratedEffectStack>();
+        m_effectStack = new AcceleratedEffectStack();
```
`adoptRef`

mattwoodrow commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -267,8 +267,15 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+        if (layerTreeHost->threadedAnimationResolutionEnabled()) {
+            LOG_WITH_STREAM(Animations, stream << "RemoteLayerTreePropertyApplier::applyProperties() found " << properties.animationChanges.effects.size() << " effects");
+            layerTreeNode->setAcceleratedEffectsAndBaseValues(properties.animationChanges.effects, properties.animationChanges.baseValues, layerTreeHost);
```
This probably doesn't matter, but since the two paths are reading different parts of `animationChanges`, maybe we should just call both?

Might help if the setting got changed between when the transaction was created and when it was received.

graouts replied:
I don't think it's worth bothering about this, at least now. For A/B testing I think it's reasonable to expect a page reload.

mattwoodrow reviewed (commented):
I'll let Simon have a look, since I wrote some of this. Looks good though!

smfr commented on Source/WebCore/platform/animation/AcceleratedEffectStack.h:
```diff
@@ -35,10 +35,10 @@ class AcceleratedEffect;
 
-class AcceleratedEffectStack {
-    WTF_MAKE_FAST_ALLOCATED;
+class AcceleratedEffectStack : public RefCounted<AcceleratedEffectStack> {
```
`WTF_MAKE_ISO_ALLOCATED`?

graouts replied:
I've added this in the newer patch.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -267,8 +267,15 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
+    if (properties.changedProperties & LayerChange::AnimationsChanged) {
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+        if (layerTreeHost->threadedAnimationResolutionEnabled()) {
+            LOG_WITH_STREAM(Animations, stream << "RemoteLayerTreePropertyApplier::applyProperties() found " << properties.animationChanges.effects.size() << " effects");
```
Log the layerID here

graouts replied:
Will be fixed in next patch.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h:
```diff
@@ -0,0 +1,47 @@
+private:
+    explicit RemoteAcceleratedEffectStack(const WebCore::FloatRect&, Seconds);
+
+    WebCore::FloatRect m_bounds;
```
Not totally clear what the "bounds" means. What coordinate system is it in?

graouts replied:
I think we can drop this for now and add it with more particulars when we get to the blending part and we make use of the bounds. Removed this in the newer patch.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -76,6 +76,8 @@ class RemoteLayerTreeDrawingAreaProxy : public DrawingAreaProxy {
     void viewWillEndLiveResize() final;
 
 #if ENABLE(THREADED_ANIMATION_RESOLUTION)
+    void animationEffectStackWasAddedToLayer(PlatformLayer*, Ref<RemoteAcceleratedEffectStack> effects);
```
Does including "EffectsStack" in the names really add anything?

graouts replied:
I think we should actually call this `effectStackWasAddedToLayer` since this is about an effect stack being added to a layer :) But maybe you want something snappier like `animationsWereAddedToLayer`?

graouts replied:
I'm going to go with `animationsWereAddedToLayer` and `animationsWereRemoved` to clarify these are animation effects.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
```diff
@@ -168,6 +174,10 @@ class RemoteLayerTreeNode : public CanMakeWeakPtr<RemoteLayerTreeNode> {
     std::optional<WebCore::RenderingResourceIdentifier> m_asyncContentsIdentifier;
+
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+    RefPtr<RemoteAcceleratedEffectStack> m_effects;
```
`m_animationEffects`? "Effects" normally describes filters etc.

graouts replied:
I guess the term "effect" is now vague indeed, but in this case `m_effectStack` would be better given the member's class name I reckon.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -238,4 +243,28 @@
+void RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues(const WebCore::AcceleratedEffects& effects, const WebCore::AcceleratedEffectValues& baseValues, RemoteLayerTreeHost* host)
+{
+    ASSERT(isUIThread());
+    ASSERT(host);
```
Can `host` just be passed by reference?

graouts replied:
Changed this in the newer patch.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -238,4 +243,28 @@
+    ASSERT(host);
+
+    if (m_effects)
+        host->animationEffectStackWasRemoved(*m_effects);
```
Should we compare old and new effects before calling this?

graouts replied:
This will _only_ be called if effects were sent over to the UI process through IPC which resulted from effects having changed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
```diff
@@ -38,6 +38,10 @@
 constexpr WebCore::FramesPerSecond DisplayLinkFramesPerSecond = 60;
 
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+#import <WebCore/ScrollingThread.h>
```
Weird that an iOS file needs to know about the macOS-only scrolling thread.

graouts replied:
In fact it does not, at least at this stage. However the `ScrollingThread` is no longer macOS-only and is guarded by `#if ENABLE(SCROLLING_THREAD) || ENABLE(THREADED_ANIMATION_RESOLUTION)`. It was my understanding your suggestion was to use this thread for this work instead of introducing a new one.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
```diff
@@ -75,6 +80,10 @@ class RemoteScrollingCoordinatorProxyIOS final : public RemoteScrollingCoordinat
 #endif
+
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+    Vector<std::pair<RetainPtr<PlatformLayer>, Ref<RemoteAcceleratedEffectStack>>> m_effectsStack;
```
I don't like retaining platform layers here. Can we store `PlatformLayerIdentifier` and look up layers on the fly?

smfr replied:
Why is this needed since RemoteLayerTreeNodes own the effects?

mattwoodrow replied:
Yeah, for iOS, I think we can do the obvious thing and just reference the Nodes.

graouts replied:
In the latest patch this class now has `HashSet<WeakPtr<RemoteLayerTreeNode>> m_animatedNodes;` to track the nodes that have animations set on them.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
```diff
@@ -160,6 +163,8 @@ class RemoteLayerTreeEventDispatcher : public ThreadSafeRefCounted<RemoteLayerTr
 
     std::unique_ptr<RunLoop::Timer> m_delayedRenderingUpdateDetectionTimer;
 
+    HashSet<Ref<RemoteAcceleratedEffectStack>> m_effects;
```
Also not sure if we should store these here. Can we just store a set of nodes that have effects that we need to pay attention to?

In general, try to avoid storing data in multiple places.

mattwoodrow replied:
The intent here is for these to be read (for animation updates) from the scrolling thread.

Hard to see without the future patches, but the intent here is that this HashSet will be made thread safe for this, whereas trying to access `RemoteLayerTreeNodes` would not be.

graouts replied:
In the latest patch this class now has `HashMap<WebCore::PlatformLayerIdentifier, Ref<RemoteAcceleratedEffectStack>> m_effectStacks;` as the unique place where the effects are stored for macOS.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
```diff
@@ -280,6 +280,16 @@
 
+void RemoteScrollingCoordinatorProxyMac::animationEffectStackWasAddedToLayer(PlatformLayer*, Ref<RemoteAcceleratedEffectStack> effects)
+{
+    m_wheelEventDispatcher->animationEffectStackWasAddedToLayer(WTFMove(effects));
```
`m_wheelEventDispatcher` is a confusing name here. Who'd have thought that a mouse wheel thing cared about animations.

graouts replied:
Renaming this to `m_eventDispatcher` in https://github.com/WebKit/WebKit/pull/21408.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -267,8 +267,15 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
+    if (properties.changedProperties & LayerChange::AnimationsChanged) {
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+        if (layerTreeHost->threadedAnimationResolutionEnabled()) {
+            LOG_WITH_STREAM(Animations, stream << "RemoteLayerTreePropertyApplier::applyProperties() found " << properties.animationChanges.effects.size() << " effects on layer " << layerTreeNode->layerID());
```
In logs I'd prefer to see the layerID first.

graouts replied:
Will bring the layer ID up in the logging message when landing.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h:
```diff
@@ -0,0 +1,47 @@
+private:
+    explicit RemoteAcceleratedEffectStack(Seconds);
+
+    Seconds m_acceleratedTimelineTimeOrigin;
```
Shouldn't an origin be `MonotonicTime`?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm:
```diff
@@ -0,0 +1,50 @@
+}
+
+RemoteAcceleratedEffectStack::RemoteAcceleratedEffectStack(Seconds acceleratedTimelineTimeOrigin)
+    : AcceleratedEffectStack()
```
This isn't necessary

graouts replied:
Will remove when landing.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
```diff
@@ -131,6 +133,11 @@ class RemoteLayerTreeNode : public CanMakeWeakPtr<RemoteLayerTreeNode> {
 
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+    void setAcceleratedEffectsAndBaseValues(const WebCore::AcceleratedEffects&, const WebCore::AcceleratedEffectValues&, RemoteLayerTreeHost&);
+    RefPtr<RemoteAcceleratedEffectStack>&& takeEffectStack() { return WTFMove(m_effectStack); }
```
`std::exchange()`

graouts replied:
Will use this when landing.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
```diff
@@ -131,6 +133,11 @@ class RemoteLayerTreeNode : public CanMakeWeakPtr<RemoteLayerTreeNode> {
     }
 
+#if ENABLE(THREADED_ANIMATION_RESOLUTION)
+    void setAcceleratedEffectsAndBaseValues(const WebCore::AcceleratedEffects&, const WebCore::AcceleratedEffectValues&, RemoteLayerTreeHost&);
```
It often makes sense for setters to take r-value references if possible.

graouts replied:
In this case we don't move the values, just copy/clone them. We may revisit this later.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
```diff
@@ -238,4 +243,27 @@
+    m_effectStack = RemoteAcceleratedEffectStack::create(host.acceleratedTimelineTimeOrigin());
+
+    auto clonedEffects = effects;
+    auto clonedBaseValues = baseValues.clone();
```
Confusing that you call `clone` on base values but not effects.

graouts replied:
Yes, this will become more clear when we get to the patch where we actually resolve animations.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -368,6 +368,18 @@
+
+void RemoteScrollingCoordinatorProxyIOS::animationsWereRemovedFromNode(RemoteLayerTreeNode& node)
+{
+    m_animatedNodeLayerIDs.remove(node.layerID());
```
Does this function mean *all* animations were removed?

graouts replied:
Yes, it does.

graouts committed:
[threaded-animation-resolution] store accelerated effects and base values in the remote layer tree
https://bugs.webkit.org/show_bug.cgi?id=265880

Reviewed by Simon Fraser.

While we correctly encode the `AcceleratedEffectStack` from WebCore to be sent over IPC to
the UI process whenever there is a meaningful change, we don't do anything with these effects
yet. This patch creates a `RemoteAcceleratedEffectStack` and store effects and base values on it
and lays the groundwork for the next patch for the threaded animation resolution feature where
we will schedule and resolve animations in the remote layer tree.

On iOS, the `RemoteAcceleratedEffectStack` is owned by a `RemoteLayerTreeNode` which will access
on the main thread for animation resolution.

On macOS, the `RemoteAcceleratedEffectStack` is owned by the `RemoteLayerTreeEventDispatcher` which
will allow access in a thread-safe manner for animation resolution.

* Source/WebCore/platform/animation/AcceleratedEffectStack.h:
(WebCore::AcceleratedEffectStack::primaryLayerEffects const): Deleted.
(WebCore::AcceleratedEffectStack::backdropLayerEffects const): Deleted.
(WebCore::AcceleratedEffectStack::baseValues): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffectStack.mm:
(WebCore::AcceleratedEffectStack::create):
(WebCore::AcceleratedEffectStack::AcceleratedEffectStack):
(WebCore::AcceleratedEffectStack::setEffects):
(WebCore::AcceleratedEffectStack::~AcceleratedEffectStack): Deleted.
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::setAcceleratedEffectsAndBaseValues):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm: Added.
(WebKit::RemoteAcceleratedEffectStack::create):
(WebKit::RemoteAcceleratedEffectStack::RemoteAcceleratedEffectStack):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationsWereAddedToNode):
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::threadedAnimationResolutionEnabled const):
(WebKit::RemoteLayerTreeHost::animationsWereAddedToNode):
(WebKit::RemoteLayerTreeHost::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::takeEffectStack):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxy::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyIOS::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::animationsWereAddedToNode):
(WebKit::RemoteLayerTreeEventDispatcher::animationsWereRemovedFromNode):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::animationsWereAddedToNode):
(WebKit::RemoteScrollingCoordinatorProxyMac::animationsWereRemovedFromNode):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Co-authored-by: Matt Woodrow <mattwoodrow@apple.com>
Canonical link: https://commits.webkit.org/271692@main
